🍌 feat: Gemini Image Generation Tool (Nano Banana)

usnavy13 opened the pull request:
## Summary

This PR adds a comprehensive Gemini Image Generation tool for LibreChat Agents with flexible authentication supporting both Gemini API and Google Cloud Vertex AI.

**Key Features:**
- **Multi-Authentication Support**: 
  - User-provided API keys (via GUI)
  - Admin-configured API keys (GEMINI_API_KEY or GOOGLE_KEY)
  - Vertex AI service accounts (automatic fallback)
- **Image Generation Capabilities**: 
  - Text-to-image generation with Gemini Nano Banana (`gemini-2.5-flash-image`) and Nano Banana Pro (`gemini-3-pro-image`)
  - Image context-aware generation (use existing images as inspiration/reference)
- **Storage Compatibility**: Works with all LibreChat file storage strategies (local, S3, Azure, Firebase)
- **Consistent Implementation**: Uses same `loadServiceKey` pattern as main Google endpoint and Anthropic Vertex AI
- **User-Friendly**: Clear safety filter messages for content policy violations

---

## Why a Tool-Based Approach?

Gemini's image generation models (`gemini-2.5-flash-image`, `gemini-3-pro-image-preview`, etc.) require special API parameters to return images:

```javascript
responseModalities: ['TEXT', 'IMAGE']
```

### Architectural Constraint

LibreChat's current chat architecture routes all endpoint requests (including Google) through the unified agents system (`/api/agents/chat/:endpoint`), which uses the `@librechat/agents` package with LangChain.

**The limitation:**
- The `@librechat/agents` package's `CustomChatGoogleGenerativeAI` class creates the Google client with a hardcoded `generationConfig` that doesn't include `responseModalities`
- LangChain's `@langchain/google-genai` package doesn't currently expose this parameter
- Without `responseModalities: ['TEXT', 'IMAGE']`, Gemini returns text descriptions instead of actual images

**Why native model selection won't work (yet):**
- Even with image models available in the model dropdown, the agents system can't request image output
- The API call succeeds but returns only text, resulting in empty responses

### The Tool Approach Works

This tool uses the `@google/genai` SDK directly (not LangChain), which:
- ✅ Supports `responseModalities: ['TEXT', 'IMAGE']`
- ✅ Returns inline image data that can be saved and displayed
- ✅ Integrates seamlessly with the agents tool system

### Future Native Support

Native support for Gemini image models as selectable endpoints would require:
1. Updates to `@librechat/agents` to support `responseModalities` in `CustomChatGoogleGenerativeAI`
2. Handling of image content parts in the response pipeline

A separate feature request has been opened for the `@librechat/agents` package.
https://github.com/danny-avila/agents/pull/41

---

## Configuration Options

```env
# Option A: Vertex AI with Service Account (recommended for production)
GEMINI_VERTEX_ENABLED=true
GOOGLE_SERVICE_KEY_FILE=/path/to/service-account.json
GOOGLE_LOC=us-central1  # or GOOGLE_CLOUD_LOCATION=global

# Option B: Dedicated Gemini API Key
GEMINI_API_KEY=your-gemini-api-key

# Option C: Shared Google API Key (uses same key as Google chat)
GOOGLE_KEY=your-google-api-key

# Optional: Change model (default: gemini-2.5-flash-image - Nano Banana)
GEMINI_IMAGE_MODEL=gemini-3-pro-image  # Nano Banana Pro
```

**Authentication Priority:**
1. User-provided API key (via GUI when adding tool)
2. `GEMINI_API_KEY` env var (admin-configured)
3. `GOOGLE_KEY` env var (shared with Google chat endpoint)
4. Vertex AI service account (automatic fallback)

This allows:
- ✅ Vertex AI users: Tool works immediately without API keys
- ✅ API key users: Admin sets global key OR users provide their own
- ✅ Mixed environments: Vertex AI with optional user override

**Related:**
- Builds upon feedback from #9538
- Uses same Vertex AI patterns as `feat/anthropic-vertex-ai` branch
- cc @devilb2103 @danny-avila

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing

Tested locally with multiple authentication configurations:

**Vertex AI Configuration:**
1. ✅ Service account authentication with `GEMINI_VERTEX_ENABLED=true`
2. ✅ Tool available without API key prompts
3. ✅ Text-to-image generation with Nano Banana (`gemini-2.5-flash-image`)
4. ✅ Image-to-image editing with context

**API Key Configuration:**
1. ✅ User-provided keys via GUI
2. ✅ Admin `GEMINI_API_KEY` env var
3. ✅ Shared `GOOGLE_KEY` env var

**General Testing:**
1. ✅ Safety filter handling for blocked content (clear error messages)
2. ✅ Local file storage strategy
3. ✅ Image context/editing with uploaded images
4. ✅ Both Nano Banana and Nano Banana Pro models

### **Test Configuration**:
- Node.js v20
- MongoDB 7.x
- Vertex AI service account (us-central1 region)
- Local file storage strategy
- Models: gemini-2.5-flash-image (Nano Banana), gemini-3-pro-image (Nano Banana Pro)

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes (updated `.env.example`)
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes
- [x] A pull request for updating the documentation has been submitted [(Here)](https://github.com/LibreChat-AI/librechat.ai/pull/452)

devilb2103 committed:
Added fully functioning Agent Tool supporting Google's Nano Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
🔧 refactor: Update Google credentials handling in GeminiImageGen.js

* Refactored the credentials path to follow a consistent pattern with other Google service integrations, allowing for an environment variable override.
* Updated documentation in README-GeminiNanoBanana.md to reflect the new credentials handling approach and removed references to hardcoded paths.

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
🛠️ refactor: Remove unnecessary whitespace in handleTools.js

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

usnavy13 committed:
🔧 feat: Update Gemini Image Generation Tool

- Bump @google/genai package version to ^1.19.0 for improved functionality.
- Refactor GeminiImageGen to createGeminiImageTool for better clarity and consistency.
- Enhance manifest.json for Gemini Image Tools with updated descriptions and icon.
- Add SVG icon for Gemini Image Tools.
- Implement progress tracking for Gemini image generation in the UI.
- Introduce new toolkit and context handling for image generation tools.

This update improves the Gemini image generation capabilities and user experience.

usnavy13 committed:
Merge remote-tracking branch 'origin/dev' into Gemini-Nano-Banana

usnavy13 committed:
🗑️ chore: Remove outdated Gemini image generation PNG and update SVG icon

- Deleted the obsolete PNG file for Gemini image generation.
- Updated the SVG icon with a new design featuring a gradient and shadow effect, enhancing visual appeal and consistency.

usnavy13 committed:
fix: ESLint formatting and unused variable in GeminiImageGen

usnavy13 committed:
fix: Update default model to gemini-2.5-flash-image

avimar commented:
This is great that it's a tool - it can be called by other models.

But yes, there's more models that can natively return text AND images (and audio?), so that would be good if it can handle that too.

usnavy13 committed:
Merge branch 'dev' into Gemini-Nano-Banana

usnavy13 committed:
✨ feat: Enhance Gemini Image Generation Configuration

- Updated .env.example to include new environment variables for Google Cloud region, service account configuration, and Gemini API key options.
- Modified GeminiImageGen.js to support both user-provided API keys and Vertex AI service accounts, improving flexibility in client initialization.
- Updated manifest.json to reflect changes in authentication methods for the Gemini Image Tools.
- Bumped @google/genai package version to 1.19.0 in package-lock.json for compatibility with new features.

paulchaum commented:
Correct me if I'm wrong, but looking at the PR, I get the impression that the tool will only work with the Gemini or Vertex AI API. I think it would be nice to have the option to make it work with any OpenAI-compatible API, such as OpenRouter.

usnavy13 committed:
Merge remote-tracking branch 'origin/dev' into Gemini-Nano-Banana

usnavy13 committed:
🔧 fix: Format Default Service Key Path in GeminiImageGen.js

- Adjusted the return statement in getDefaultServiceKeyPath function for improved readability by formatting it across multiple lines. This change enhances code clarity without altering functionality.

usnavy13 committed:
Merge remote-tracking branch 'origin/dev' into Gemini-Nano-Banana

usnavy13 committed:
✨ feat: Enhance Gemini Image Generation with Token Usage Tracking

- Added `recordTokenUsage` function to track token usage for balance management.
- Integrated token recording into the image generation process.
- Updated Gemini image generation tool to accept optional `aspectRatio` and `imageSize` parameters for improved image customization.
- Updated token values for new Gemini models in the transaction model.
- Improved documentation for image generation tool descriptions and parameters.

usnavy13 committed:
✨ feat: Add new Gemini models for image generation token limits

- Introduced token limits for 'gemini-3-pro-image' and 'gemini-2.5-flash-image' models.
- Updated token values to enhance the Gemini image generation capabilities.

usnavy13 committed:
Merge origin/dev into Gemini-Nano-Banana

Resolved conflicts:
- api/package.json: Keep @google/genai (new SDK), accept dev changes
- package-lock.json: Regenerated with npm install

usnavy13 committed:
🔧 fix: Update Google Service Key Path for Consistency in Initialization (#11001)

usnavy13 committed:
Merge branch 'dev' into Gemini-Nano-Banana

usnavy13 committed:
Merge remote-tracking branch 'origin/dev' into Gemini-Nano-Banana

usnavy13 committed:
Merge origin/dev into Gemini-Nano-Banana

Resolved conflicts in peerDependencies by keeping both:
- @google/genai (from feature branch)
- @aws-sdk/client-bedrock-runtime (from dev)

Also merged transitive dependencies in package-lock.json.

inv-Eldho commented:
@danny-avila 
Hi, 
Could you please move forward with this pull request? We’re really looking forward to this feature

danny-avila committed:
🔧 refactor: Update GeminiImageGen for improved file handling and path resolution

- Changed the default service key path to use process.cwd() for better compatibility.
- Replaced synchronous file system operations with asynchronous promises for mkdir and writeFile, enhancing performance and error handling.
- Added error handling for credential file access to prevent crashes when the file does not exist.

danny-avila committed:
🔧 refactor: Update GeminiImageGen to streamline API key handling

- Refactored API key checks to improve clarity and consistency.
- Removed redundant checks for user-provided keys, enhancing code readability.
- Ensured proper logging for API key usage across different configurations.

danny-avila committed:
🔧 fix: Update GeminiImageGen to handle imageSize support conditionally

- Added a check to ensure imageSize is only applied if the gemini model does not include 'gemini-2.5-flash-image', improving compatibility.
- Enhanced the logic for setting imageConfig to prevent potential issues with unsupported configurations.

danny-avila committed:
🔧 refactor: Simplify local storage condition in createGeminiImageTool function

danny-avila committed:
🔧 feat: Enhance image format handling in GeminiImageGen with conversion support

danny-avila committed:
🔧 refactor: Streamline API key initialization in GeminiImageGen

- Simplified the handling of API keys by removing redundant checks for user-provided keys.
- Updated logging to reflect the new priority order for API key usage, enhancing clarity and consistency.
- Improved code readability by consolidating key retrieval logic.

danny-avila commented:
Made a few necessary changes to get this ready for merge:

1. **`imageSize` parameter**: Only valid for `gemini-3-pro-image-preview`, not `gemini-2.5-flash-image` (undocumented by Google but causes API errors). Added conditional check.

2. **Format handling**: Pro-preview wasn't outputting PNG as hardcoded. I think it returns either JPG or WEBP. Also, the `librechat.yaml` `imageOutputType` config wasn't being respected like it is for OpenAI tools. Fixed to detect actual MIME type from response.

3. **API key refactoring**: Auth values are already resolved by the tool loader, there's no need to check if something is user provided or not.

4. **Path/async operations**: Updated service key path to match main Google endpoint + converted to async file ops.

I was hesitant to merge this since LC is moving away from native tools (especially AI provider API wrappers) in favor of MCP, but I am only merging given the amount of work already put into this by others. Maintaining this tool will not be high priority.

KiGamji commented:
> Made a few necessary changes to get this ready for merge:
> 
>     1. **`imageSize` parameter**: Only valid for `gemini-3-pro-image-preview`, not `gemini-2.5-flash-image` (undocumented by Google but causes API errors). Added conditional check.
> 
>     2. **Format handling**: Pro-preview wasn't outputting PNG as hardcoded. I think it returns either JPG or WEBP. Also, the `librechat.yaml` `imageOutputType` config wasn't being respected like it is for OpenAI tools. Fixed to detect actual MIME type from response.
> 
>     3. **API key refactoring**: Auth values are already resolved by the tool loader, there's no need to check if something is user provided or not.
> 
>     4. **Path/async operations**: Updated service key path to match main Google endpoint + converted to async file ops.
> 
> 
> I was hesitant to merge this since LC is moving away from native tools (especially AI provider API wrappers) in favor of MCP, but I am only merging given the amount of work already put into this by others. Maintaining this tool will not be high priority.

Do you have any plans for multi-modal image generation?

KiGamji commented:
Appreciate it man!

RedwindA commented:
Does it respect GOOGLE_REVERSE_PROXY in .env?

usnavy13 commented:
> Why it always generate multiple images in response to a single user request?  Even if we ask it to only generate one image per one request in the agent's prompt, it still generates multiple images.


I do not have this issue

teecrow commented:
> Why it always generate multiple images in response to a single user request? Even if we ask it to only generate one image per one request in the agent's prompt, it still generates multiple images.

This also happened to me when, in the configuration for the Agent, I set the "Model" to `gemini-3-flash-preview`, but it resolved when I changed the model to `gemini-3-pro-preview`.
